File: android/app/src/main/java/org/cliprelay/tcp/TcpImageReceiver.kt
package org.cliprelay.tcp

import java.io.InputStream
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket
import java.net.SocketTimeoutException
import java.util.concurrent.atomic.AtomicBoolean

data class ServerInfo(val host: String, val port: Int)

class TcpImageReceiver(
    private val expectedSize: Int,
    private val allowedSenderIp: String?,
    private val noConnectionTimeoutMs: Int = 30_000,
    private val transferTimeoutMs: Int = 120_000,
    private val maxConnections: Int = 2,
) {
    private val cancelled = AtomicBoolean(false)
    private var serverSocket: ServerSocket? = null

    fun start(): ServerInfo {
        val server = ServerSocket()
        server.reuseAddress = true
        server.bind(InetSocketAddress("0.0.0.0", 0))
        serverSocket = server

        val localIp = NetworkUtil.getLocalIpAddress() ?: "0.0.0.0"
        return ServerInfo(localIp, server.localPort)
    }

    /**
     * Blocks until [expectedSize] bytes are received from an allowed sender,
     * or throws [TcpTransferException] on error/timeout/cancel.
     */
    fun receive(): ByteArray {
        val server = serverSocket ?: throw TcpTransferException("Server not started")

        try {
            server.soTimeout = noConnectionTimeoutMs

            var attemptsLeft = maxConnections
            while (attemptsLeft > 0 && !cancelled.get()) {
                attemptsLeft--

                val client: Socket
                try {
                    client = server.accept()
                } catch (e: SocketTimeoutException) {
                    throw TcpTransferException("No connection within ${noConnectionTimeoutMs}ms", e)
                }

                try {
                    val remoteIp = (client.remoteSocketAddress as? InetSocketAddress)
                        ?.address?.hostAddress

                    if (allowedSenderIp != null && remoteIp != allowedSenderIp) {
                        client.close()
                        continue
                    }

                    client.soTimeout = transferTimeoutMs
                    val data = readExactly(client.getInputStream(), expectedSize)
                    return data
                } finally {
                    try { client.close() } catch (_: Exception) {}
                }
            }

            throw TcpTransferException("No valid connection after $maxConnections attempts")
        } catch (e: TcpTransferException) {
            throw e
        } catch (e: Exception) {
            if (cancelled.get()) {
                throw TcpTransferException("Transfer cancelled")
            }
            throw TcpTransferException("Receive failed: ${e.message}", e)
        }
    }

    fun cancel() {
        cancelled.set(true)
        close()
    }

    fun close() {
        try { serverSocket?.close() } catch (_: Exception) {}
    }

    private fun readExactly(input: InputStream, size: Int): ByteArray {
        val buffer = ByteArray(size)
        var offset = 0
        while (offset < size) {
            if (cancelled.get()) throw TcpTransferException("Transfer cancelled")
            val read = input.read(buffer, offset, size - offset)
            if (read == -1) {
                throw TcpTransferException(
                    "Stream closed after $offset bytes, expected $size"
                )
            }
            offset += read
        }
        return buffer
    }
}

File: macos/ClipRelayMac/Sources/TCP/TcpImageReceiver.swift
import Foundation

struct TcpServerInfo {
    let host: String
    let port: UInt16
}

final class TcpImageReceiver {
    private let expectedSize: Int
    private let allowedSenderIp: String?
    private let noConnectionTimeoutMs: Int
    private let transferTimeoutMs: Int
    private let maxConnections: Int

    private var serverFd: Int32 = -1
    private var cancelled = false
    private let lock = NSLock()

    init(
        expectedSize: Int,
        allowedSenderIp: String?,
        noConnectionTimeoutMs: Int = 30_000,
        transferTimeoutMs: Int = 120_000,
        maxConnections: Int = 2
    ) {
        self.expectedSize = expectedSize
        self.allowedSenderIp = allowedSenderIp
        self.noConnectionTimeoutMs = noConnectionTimeoutMs
        self.transferTimeoutMs = transferTimeoutMs
        self.maxConnections = maxConnections
    }

    func start() throws -> TcpServerInfo {
        serverFd = socket(AF_INET, SOCK_STREAM, 0)
        guard serverFd >= 0 else {
            throw TcpTransferError.receiveFailed("socket() failed: \(errno)")
        }

        var yes: Int32 = 1
        setsockopt(serverFd, SOL_SOCKET, SO_REUSEADDR, &yes, socklen_t(MemoryLayout<Int32>.size))

        var addr = sockaddr_in()
        addr.sin_len = UInt8(MemoryLayout<sockaddr_in>.size)
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port = 0 // OS-assigned
        addr.sin_addr.s_addr = INADDR_ANY.bigEndian

        let bindResult = withUnsafePointer(to: &addr) { ptr in
            ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sa in
                bind(serverFd, sa, socklen_t(MemoryLayout<sockaddr_in>.size))
            }
        }
        guard bindResult == 0 else {
            close(serverFd)
            throw TcpTransferError.receiveFailed("bind() failed: \(errno)")
        }

        guard listen(serverFd, 2) == 0 else {
            close(serverFd)
            throw TcpTransferError.receiveFailed("listen() failed: \(errno)")
        }

        // Get assigned port
        var boundAddr = sockaddr_in()
        var addrLen = socklen_t(MemoryLayout<sockaddr_in>.size)
        withUnsafeMutablePointer(to: &boundAddr) { ptr in
            ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sa in
                _ = getsockname(serverFd, sa, &addrLen)
            }
        }
        let port = UInt16(bigEndian: boundAddr.sin_port)

        let localIp = LocalNetworkAddress.getLocalIPv4Address() ?? "0.0.0.0"
        return TcpServerInfo(host: localIp, port: port)
    }

    func receive() throws -> Data {
        guard serverFd >= 0 else {
            throw TcpTransferError.serverNotStarted
        }

        var attemptsLeft = maxConnections
        while attemptsLeft > 0 {
            lock.lock()
            let isCancelled = cancelled
            lock.unlock()
            if isCancelled { throw TcpTransferError.transferCancelled }

            attemptsLeft -= 1

            // Use poll() for accept timeout since SO_RCVTIMEO doesn't affect accept() on macOS
            var pfd = pollfd(fd: serverFd, events: Int16(POLLIN), revents: 0)
            let pollResult = poll(&pfd, 1, Int32(noConnectionTimeoutMs))
            if pollResult == 0 {
                throw TcpTransferError.noConnection(timeoutMs: noConnectionTimeoutMs)
            }
            if pollResult < 0 {
                lock.lock()
                let wasCancelled = cancelled
                lock.unlock()
                if wasCancelled { throw TcpTransferError.transferCancelled }
                throw TcpTransferError.receiveFailed("poll() failed: \(errno)")
            }

            var clientAddr = sockaddr_in()
            var clientAddrLen = socklen_t(MemoryLayout<sockaddr_in>.size)
            let clientFd = withUnsafeMutablePointer(to: &clientAddr) { ptr in
                ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sa in
                    accept(serverFd, sa, &clientAddrLen)
                }
            }

            if clientFd < 0 {
                lock.lock()
                let wasCancelled = cancelled
                lock.unlock()
                if wasCancelled { throw TcpTransferError.transferCancelled }
                throw TcpTransferError.receiveFailed("accept() failed: \(errno)")
            }

            // Validate sender IP
            if let allowed = allowedSenderIp {
                var remoteHostname = [CChar](repeating: 0, count: Int(NI_MAXHOST))
                withUnsafePointer(to: &clientAddr) { ptr in
                    ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sa in
                        _ = getnameinfo(sa, clientAddrLen, &remoteHostname,
                                        socklen_t(remoteHostname.count), nil, 0, NI_NUMERICHOST)
                    }
                }
                let remoteIp = String(cString: remoteHostname)
                if remoteIp != allowed {
                    close(clientFd)
                    continue
                }
            }

            // Set transfer timeout
            setReceiveTimeout(fd: clientFd, ms: transferTimeoutMs)

            do {
                let data = try readExactly(fd: clientFd, size: expectedSize)
                close(clientFd)
                return data
            } catch {
                close(clientFd)
                throw error
            }
        }

        throw TcpTransferError.noValidConnection(maxAttempts: maxConnections)
    }

    func cancel() {
        lock.lock()
        cancelled = true
        lock.unlock()
        closeServer()
    }

    func closeServer() {
        if serverFd >= 0 {
            close(serverFd)
            serverFd = -1
        }
    }

    // MARK: - Private

    private func setReceiveTimeout(fd: Int32, ms: Int) {
        var tv = timeval()
        tv.tv_sec = ms / 1000
        tv.tv_usec = Int32((ms % 1000) * 1000)
        setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &tv, socklen_t(MemoryLayout<timeval>.size))
    }

    private func readExactly(fd: Int32, size: Int) throws -> Data {
        var buffer = Data(count: size)
        var offset = 0
        while offset < size {
            lock.lock()
            let isCancelled = cancelled
            lock.unlock()
            if isCancelled { throw TcpTransferError.transferCancelled }

            let n = buffer.withUnsafeMutableBytes { rawPtr in
                let ptr = rawPtr.baseAddress!.advanced(by: offset)
                return read(fd, ptr, size - offset)
            }
            if n == 0 {
                throw TcpTransferError.streamClosed(received: offset, expected: size)
            }
            if n < 0 {
                if errno == EAGAIN || errno == EWOULDBLOCK {
                    throw TcpTransferError.receiveFailed("Transfer timed out after \(transferTimeoutMs)ms")
                }
                throw TcpTransferError.receiveFailed("read() failed: \(errno)")
            }
            offset += n
        }
        return buffer
    }
}
